src/lib/register-daemon.js

```diff
@@ -1,29 +1,51 @@
 import { store, createDaemon, logger } from '../utils'
 import { app } from 'electron'
+import fs from 'fs-extra'
+import { join } from 'path'
 
 export default async function (ctx) {
   let config = store.get('ipfsConfig')
-  ctx.ipfsd = await createDaemon(config)
-
-  // Update the path if it was blank previously.
-  // This way we use the default path when it is
-  // not set.
-  if (config.path === '') {
-    config.path = ctx.ipfsd.repoPath
-    store.set('ipfsConfig', config)
+  let ipfsd = null
+
+  ctx.getIpfsd = () => ipfsd
+
+  ctx.startIpfs = async () => {
+    ipfsd = await createDaemon(config)
+
+    // Update the path if it was blank previously.
+    // This way we use the default path when it is
+    // not set.
+    if (config.path === '') {
+      config.path = ipfsd.repoPath
+      store.set('ipfsConfig', config)
+    }
   }
 
-  app.once('will-quit', async (event) => {
-    event.preventDefault()
-    logger.info('Stopping daemon')
+  ctx.stopIpfs = async () => {
+    if (!fs.pathExists(join(ipfsd.repoPath, 'config'))) {
+      // Is remote api... ignore
+      ipfsd = null
+      return
+    }
 
-    await new Promise((resolve, reject) => {
-      ctx.ipfsd.stop(err => {
-        if (err) reject(err)
-        else resolve()
+    return new Promise((resolve, reject) => {
+      ipfsd.stop(err => {
+        if (err) {
+          return reject(err)
+        }
+
+        ipfsd = null
+        resolve()
       })
     })
+  }
+
+  await ctx.startIpfs()
 
+  app.once('will-quit', async (event) => {
+    event.preventDefault()
```

Comment:
@hacdias Fix from https://github.com/ipfs-shipyard/ipfs-desktop/pull/779/commits/88b586cf0e84b22b3b3b0a196848e4428ff915ad did not help on linux, however if I remove this line `Quit` action starts working as expected  (with both embedded and remote nodes):

```suggestion
```


I think we were just breaking shutdown flow:
> Calling `event.preventDefault()` will prevent the default behaviour, which is terminating the application.
> https://github.com/electron/electron/blob/master/docs/api/app.md#event-will-quit

Is  `event.preventDefault()` necessary? My intuition is that shutdown will happen after `quit` event which happens after `will-quit` anyway.

Reply:
@lidel, yes, it is needed otherwise ipfsd.stop might never be called. If you notice, that's a **once** listener and in the end we call **app.quit** again so there shouldn't be any problems unless the app.quit is never reached.